File: Task_1/Square_Root/sqrt_binary_search.c
#include <stdio.h>

double sqrt_binary_search(double x, int max_iter ) {
    if (x < 0.0) {
        return -1.0;  // Error case, like in Verilog
    }

    if (x == 0.0 || x == 1.0) {
        return x;  // sqrt(0) = 0, sqrt(1) = 1
    }

    double low = 0.0;
    double high = (x >= 1.0) ? x : 1.0;
    double mid = 0.0;
    double epsilon = 1e-6;
    int iter = 0;

    while ((high - low) > epsilon && iter < max_iter) {
        mid = (low + high) / 2.0;

        if (mid * mid > x)
            high = mid;
        else
            low = mid;

        iter++;
    }

    return (low + high) / 2.0;
}

int main() {
    double x;

    printf("Enter a number to find its square root: ");
    scanf("%lf", &x);

    double result = sqrt_binary_search(x, 100);

    if (result == -1.0)
        printf("Error: Cannot compute sqrt of a negative number.\n");
    else
        printf("Approximate sqrt(%.6f) = %.6f\n", x, result);

    return 0;
}

File: Task_1/Square_Root/sqrt_binary_search.py
def sqrt(x, max_iter=100 ):
    if x < 0:
        raise ValueError("Invalid input ; neg num")

    if x == 0 or x == 1:
        return x


    low = 0.0
    high = x if x >= 1.0 else 1.0
    iter = 0

    while (high - low) > 1e-6 and iter < max_iter:
        mid = (low + high) / 2.0
        square = mid * mid
        if abs(square - x) <= 1e-6:
            return mid
        elif square < x:
            low = mid
        else:
            high = mid
        iter += 1

    return mid



print(sqrt(25))     
print(sqrt(16))
#print(sqrt(-45))     



